File: src/main/java/jk/kamoru/flayground/info/service/VideoInfoService.java
package jk.kamoru.flayground.info.service;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import jk.kamoru.flayground.flay.FlayNotfoundException;
import jk.kamoru.flayground.flay.domain.Flay;
import jk.kamoru.flayground.flay.service.FlayService;
import jk.kamoru.flayground.history.domain.History;
import jk.kamoru.flayground.history.service.HistoryService;
import jk.kamoru.flayground.info.domain.Tag;
import jk.kamoru.flayground.info.domain.Video;

@Service
public class VideoInfoService extends InfoServiceAdapter<Video, String> {

  @Autowired
  FlayService flayService;
  @Autowired
  HistoryService historyService;

  /**
   * 신규 video 저장. 기존에 있는지 찾아보고 동작
   *
   * @param video
   * @return
   */
  public Video put(Video putVideo) {
    Video video = this.infoSource.getOrNew(putVideo.getOpus());
    video.setTitle(putVideo.getTitle());
    video.setDesc(putVideo.getDesc());
    this.update(video);
    return video;
  }

  @Override
  public void update(Video updateVideo) {
    updateVideo.setLastAccess(new Date().getTime());
    try {
      Flay flay = flayService.get(updateVideo.getOpus());
      flay.setVideo(updateVideo);
      historyService.save(History.Action.UPDATE, flay, null);
    } catch (FlayNotfoundException ignore) {
    }
    super.update(updateVideo);
  }

  public void removeTag(Tag deleteTag) {
    List<Video> videoList = new ArrayList<>();
    for (Video video : list()) {
      if (video.getTags().contains(deleteTag)) {
        videoList.add(video);
      }
    }
    for (Video video : videoList) {
      video.getTags().remove(deleteTag);
      update(video);
    }
  }

  public void setRank(String opus, int rank) {
    Video video = this.infoSource.get(opus);
    video.setRank(rank);
    this.update(video);
  }

  public void setLike(String opus) {
    Video video = this.infoSource.get(opus);
    Date today = new Date();
    if (video.getLikes() == null || video.getLikes().stream().filter((date) -> isSameDay(today, date)).count() == 0) {
      video.addLike();
    }
    this.update(video);
  }

  private static boolean isSameDay(Date d1, Date d2) {
    Calendar c1 = Calendar.getInstance();
    c1.setTime(d1);
    Calendar c2 = Calendar.getInstance();
    c2.setTime(d2);
    return c1.get(Calendar.YEAR) == c1.get(Calendar.YEAR) && c1.get(Calendar.MONTH) == c1.get(Calendar.MONTH) && c1.get(Calendar.DAY_OF_MONTH) == c1.get(Calendar.DAY_OF_MONTH);
  }

  public void toggleTag(String opus, Tag tag, boolean checked) {
    Video video = this.infoSource.get(opus);
    if (checked) {
      video.getTags().add(tag);
    } else {
      video.getTags().remove(tag);
    }
    this.update(video);
  }

  public void setComment(String opus, String comment) {
    Video video = this.infoSource.get(opus);
    video.setComment(comment.trim());
    this.update(video);
  }

}

File: server_node/src/services/video-info.service.ts
import { createHistory } from '../domain/history';
import { Tag } from '../domain/tag';
import { Video } from '../domain/video';
import { getInstanceFlay } from '../sources/flay-source';
import { historyRepository } from '../sources/history-repository';
import { videoInfoSource } from '../sources/info-sources';
import { sseSend } from './sse-emitters';

/**
 * Video 정보 서비스.
 * Java VideoInfoService 대응
 */

/** opus로 Video 조회 */
export function get(opus: string): Video {
  return videoInfoSource.get(opus);
}

/** 전체 Video 목록 */
export function list(): Video[] {
  return videoInfoSource.getList();
}

/** query 문자열로 검색 */
export function find(query: string): Video[] {
  return videoInfoSource.getList().filter((v) => JSON.stringify(v).includes(query));
}

/** 새 Video 생성 */
export function create(video: Video): Video {
  const created = videoInfoSource.create(video);
  sseSend(created);
  return created;
}

/**
 * Video를 저장한다. 없으면 생성, 있으면 수정.
 * Java VideoInfoService.put() 대응
 */
export function put(putVideo: Video): Video {
  const video = videoInfoSource.getOrNew(putVideo.opus);
  video.title = putVideo.title;
  video.desc = putVideo.desc;
  update(video);
  return video;
}

/**
 * Video를 업데이트한다.
 * lastAccess를 갱신하고 History를 기록한다.
 * Java VideoInfoService.update() 대응
 */
export function update(updateVideo: Video): void {
  updateVideo.lastAccess = Date.now();

  // Flay가 있으면 History 기록
  try {
    const flay = getInstanceFlay(updateVideo.opus);
    flay.video = updateVideo;
    historyRepository.save(createHistory(updateVideo.opus, 'UPDATE', JSON.stringify(updateVideo)));
  } catch {
    // Instance에 없으면 무시
  }

  videoInfoSource.update(updateVideo);
  sseSend(updateVideo);
}

/** Video 삭제 */
export function deleteVideo(video: Video): void {
  videoInfoSource.delete(video.opus);
  sseSend(video);
}

/**
 * rank를 설정한다.
 * Java VideoInfoService.setRank() 대응
 */
export function setRank(opus: string, rank: number): void {
  const video = videoInfoSource.get(opus);
  video.rank = rank;
  update(video);
}

/**
 * like를 추가한다. 6시간 이내 중복 방지.
 * Java VideoInfoService.setLike() 대응
 */
export function setLike(opus: string): void {
  const video = videoInfoSource.get(opus);
  const now = Date.now();
  const SIX_HOURS = 6 * 60 * 60 * 1000;

  // likes가 없거나, 6시간 이내 like가 없으면 추가
  if (!video.likes || video.likes.filter((dateStr) => new Date(dateStr).getTime() + SIX_HOURS > now).length === 0) {
    if (!video.likes) video.likes = [];
    video.likes.push(new Date().toISOString());
    update(video);
  }
}

/**
 * 태그를 토글한다.
 * Java VideoInfoService.toggleTag() 대응
 */
export function toggleTag(opus: string, tag: Tag, checked: boolean): void {
  const video = videoInfoSource.get(opus);

  if (checked) {
    // 이미 있는지 확인
    if (!video.tags.some((t) => t.id === tag.id)) {
      video.tags.push(tag);
    }
  } else {
    video.tags = video.tags.filter((t) => t.id !== tag.id);
  }
  update(video);
}

/**
 * 코멘트를 설정한다.
 */
export function setComment(opus: string, comment: string): void {
  const video = videoInfoSource.get(opus);
  video.comment = comment.trim();
  update(video);
}

/**
 * 특정 태그를 모든 Video에서 제거한다.
 * Java VideoInfoService.removeTag() 대응
 */
export function removeTag(deleteTag: Tag): void {
  for (const video of videoInfoSource.getList()) {
    if (video.tags.some((t) => t.id === deleteTag.id)) {
      video.tags = video.tags.filter((t) => t.id !== deleteTag.id);
      update(video);
    }
  }
}
